### PR TITLE
Fix #6027: Console fails to display all output

### LINKED ISSUE #6027: console does not display all items
**OS:** 8.1
**Version:** 0.1.0
**Commit/Build:** 4af64bb

With the park linked below, entering `rides list` results in this mess:

image

https://www.nedesigns.com/park/1641/calypso-theme-park-la/

### PR DESCRIPTION
**Pull request by rwjuk**
This PR switches the console buffer from a C array to a `std::vector`, removing the need for a fixed buffer size. This way the console can display as much output as required.

I've also standardised the caret drawing code based on that in `TextInput.cpp`.

### CODE REVIEW

File: src/openrct2/interface/Console.cpp [hunk 1]
```diff
@@ -72,9 +74,11 @@ bool gConsoleOpen = false;
 static bool _consoleInitialised = false;
 static sint32 _consoleLeft, _consoleTop, _consoleRight, _consoleBottom;
 static sint32 _lastMainViewportX, _lastMainViewportY;
-static utf8 _consoleBuffer[CONSOLE_BUFFER_SIZE] = { 0 };
-static utf8 *_consoleBufferPointer = _consoleBuffer;
-static utf8 *_consoleViewBufferStart = _consoleBuffer;
+
+static utf8 * _consoleBuffer;
+static utf8 * _consoleBufferPointer;
```
Comment (IntelOrca) [inline src/openrct2/interface/Console.cpp]: I think you can remove this. It isn't read anywhere.

File: src/openrct2/interface/Console.cpp [hunk 2]
```diff
@@ -382,16 +400,9 @@ static void console_write_prompt()
...
-    size_t bufferShift = charactersToWrite - charactersRemainingInBuffer;
-    if (charactersToWrite > charactersRemainingInBuffer) {
-        memmove(_consoleBuffer, _consoleBuffer + bufferShift, CONSOLE_BUFFER_SIZE - bufferShift);
-        _consoleBufferPointer -= bufferShift;
-        charactersRemainingInBuffer = CONSOLE_BUFFER_SIZE - (_consoleBufferPointer - _consoleBuffer) - 1;
-    }
-    safe_strcpy(_consoleBufferPointer, src, charactersRemainingInBuffer);
-    _consoleBufferPointer += charactersToWrite;
+    std::string srcString(src, strlen(src));
+    _consoleVector.insert(_consoleVector.end(), srcString.begin(), srcString.end());
```
Comment (AaronVanGeffen) [inline src/openrct2/interface/Console.cpp]: You could also just use `_consoleVector.push_back()` instead to insert at the end.

Comment (AaronVanGeffen) [reply src/openrct2/interface/Console.cpp]: This line is only adding one string, though, right? The calls to `.begin()` and `.end()` on one `std::string` are a bit misleading in an iterator context, in that respect.

Comment (IntelOrca) [reply src/openrct2/interface/Console.cpp]: `_consoleVector` is a buffer not a list of strings. It needs to copy the chars from the string to the end of the buffer.

Comment (rwjuk) [reply src/openrct2/interface/Console.cpp]: Yeah, I did try using push_back() but that will only insert the first character of the string. I'll add a comment explaining this though, it isn't obvious.

Comment (AaronVanGeffen) [reply src/openrct2/interface/Console.cpp]: Hmm. Any particular reason why it's not implemented on a line basis? (Or indeed, why we're sticking with it?)

Comment (rwjuk) [reply src/openrct2/interface/Console.cpp]: Not to my knowledge - I can & will look into that, and removing some of the other C-flavour stuff. For now though this is a mergeable fix, and a step in the right direction at least!

File: src/openrct2/interface/Console.cpp [hunk 3]
```diff
@@ -298,8 +314,8 @@ void console_draw(rct_drawpixelinfo *dpi)
         sint32 caretX = x + gfx_get_string_width(lineBuffer);
         sint32 caretY = y + lineHeight;
 
-        uint8 caretColour = ColourMapA[BASE_COLOUR(backgroundColour)].lightest;
-        gfx_fill_rect(dpi, caretX, caretY, caretX + CONSOLE_CARET_WIDTH, caretY + 1, caretColour);
+        uint8 caretColour = ColourMapA[BASE_COLOUR(textColour)].lightest;
+        gfx_fill_rect(dpi, caretX, y + 9, caretX + CONSOLE_CARET_WIDTH, caretY, caretColour);
```
Comment (AaronVanGeffen) [inline src/openrct2/interface/Console.cpp]: The caret is still different compared to the text input window. Most notably, the caret is now quite thick when using e.g. Japanese — roughly twice as high as when using the sprite font. In the text input window, it is still only 1px high.

Comment (rwjuk) [reply src/openrct2/interface/Console.cpp]: That's because in all my `vector`-related excitement I screwed up the caret function! Fixed.

**Comment by rwjuk** [comment]
This is probably now superseded by #7064.

**Comment by AaronVanGeffen** [comment]
> We can cherry pick the first and third commits - once we're sure 7064 is the way to go I'll get rid of the second.

It's unfortunate that these PRs are both tackling the same thing, but that sounds like the sensible solution, yeah.

**Comment by AaronVanGeffen** [comment]
@rwjuk I have merged https://github.com/OpenRCT2/OpenRCT2/pull/7064. Could you rebase this PR, or open a new one for the cursor changes?

**Comment by rwjuk** [comment]
@AaronVanGeffen Done, see #7080.

**Pull request closed without merging**